### PR TITLE
Add options for testing and continuous deployment with GitLab CI.

### PR DESCRIPTION
**Pull request by calculuscowboy**
This is still very much a work in progress but I thought I should start a PR to get some feedback. The documentation needs work and I haven't tested it yet but just want to make sure I'm going in the right direction implementing #607. :)

The GitLab configuration is based on one I currently use in a project.  Enabling GitLab CI sets up .gitlab-ci.yml to run the tests.  There are additional options to specify branches to deploy in both staging and production environments.

This is my first substantial contribution to Cookiecutter Django so any feedback is much appreciated!

### CODE REVIEW

File: cookiecutter.json [hunk 1]
```diff
@@ -20,5 +20,9 @@
     "use_compressor": "n",
     "js_task_runner": ["Gulp", "Grunt", "None"],
     "use_lets_encrypt": "n",
+    "use_gitlab_ci": "n",
+    "staging_branch": "",
+    "staging_domain_name": "beta.{{ cookiecutter.domain_name }}",
+    "production_branch": "",
```
Comment (pydanny) [inline cookiecutter.json]: Adding 4 questions is a lot, especially for anyone unfamiliar with GitLab's system (most of our users). That's going to add a huge burden of maintenance to the core contributors.

I would much rather have a single question (`use_gitlab_ci`) in `cookiecutter.json`, then see more questions asked during a `post-generate` hook.

Comment (calculuscowboy) [inline cookiecutter.json]: How about if the option for use_gitlab_ci was "No", "Test Only" or "Test and Deploy"?  If the user chooses "Test and Deploy" then the master branch is automatically deployed to {{ cookiecutter.domain_name }}.

That would make it easier to get started with continuous deployment.  Someone wanting a more complex deployment configuration (different branch, staging site, etc) can modify .gitlab-ci.yml to their needs after the project in generated.

Comment (pydanny) [inline cookiecutter.json]: Sounds good to me!

**Comment by jayfk** [comment]
@calculuscowboy what's the status here? Is this ready to merge?

**Comment by calculuscowboy** [comment]
I think it's ready.  Do you want me to resolve the conflicts?

**Comment by jayfk** [comment]
> I think it's ready. Do you want me to resolve the conflicts?

That'd be terrific.

**Comment by ssteinerx** [comment]
This runs side-by-side with github/travix/tox?

**Comment by calculuscowboy** [comment]
@ssteinerx Yes they can co-exist.  If your project is hosted with GitHub and you're using Travis for CI, the GitLab configuration file is ignored.  If your project is in GitLab, the Travis configuration is ignored.

**Comment by calculuscowboy** [comment]
Hmm, I'm not sure what happened with the tests..

File: {{cookiecutter.project_slug}}/compose/django/Dockerfile-test [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{% if cookiecutter.use_python2 == 'n' -%}
```
Comment (ameistad) [inline {{cookiecutter.project_slug}}/compose/django/Dockerfile-test]: This line should be `{% if cookiecutter.use_python3 == 'y' -%}`

**Comment by ssteinerx** [comment]
@calculuscowboy How can this test be tested?  

Can we make github and gitlab tests both be run on checking on GitHub?

IOW, if we mirrored this repo to gitlab (must be automated!), could we make it so that every push to GitHub forces a push to GitLab and runs both sets of tests?

Is there a badge we could put in the README.md to show GitLab tests also passing?

I'm hugely in favor of this if we can get all of the above to work...

**Comment by thaxy** [comment]
+1

**Comment by thaxy** [comment]
I've managed to make my tests with GitLab CI . Now I wan't to add staging and production deployment. As far as I see it the runner basically notifies the e.g. staging server to run the recent built containers.
GitLab has a private [container registry](https://about.gitlab.com/2016/05/23/gitlab-container-registry/). Wouldn't it be better to add the successful built containers here and to make the staging server pull the most recent version?

**Comment by calculuscowboy** [comment]
@ssteinerx GitLab CI runs the same set of tests so I'm not sure there's a benefit to running both.  But if you wanted to do that, GitLab has repository mirroring.  I'm not sure what tools GitHub provides for mirroring external repositories.

There is a badge that can be put in the readme - I'll look at adding it.

@thaxy At the time I started on this, GitLab didn't have a container registry but that's another solution and would need something on the staging/production side to pull the image.

**Comment by webyneter** [comment]
@calculuscowboy, there are some conflicts out there -- mind resolving 'em?

**Comment by jayfk** [comment]
I'm going to close this for now. Once the conflicts are resolved, please feel free to submit a new pull request.

**Pull request closed without merging**